### PR TITLE
Update the Callout Component in AR

### PR DESCRIPTION
**Pull request by sophie-macmillan**
## What does this change?

Previously there was a callout component in Apps Rendering that wasn't used. It looked similar to the Callout Component in DCR as below.

The new version of the Callout Component replaces the old unused one. It is hydrated on the client side as it relies on JS (unlike the other one) and uses the new model for Callouts that has been implemented as part of this work.

## Why?
Editorial Experience is working on Callouts as part of their community focus this quarter. The old Callout wasn't rendered in Apps Rendering due to a filter. The updated callout is inline with the current UX designs.

## Screenshots
**Before**
The Callout wasn't actually used anywhere, but looked like this
image

**After**
The updated callout can be collapsable, or non collapsable, so it can be used in a standalone callout article, or embedded as part of another article.

_Collapsed Form_
image

_Successful Submission_
image

_Validation Errors_
image
_Closed Form_

image

_Video Demo_

https://user-images.githubusercontent.com/26366706/207290810-31078ec6-f987-4bbd-82f9-1ecacf24a2e2.mov

### CODE REVIEW

**Review by tkgnm — CHANGES_REQUESTED** [review]
Amazing stuff - looking really good! Some visual/UX bugs to address I believe but overall thanks for making this so easy to read.

File: apps-rendering/src/components/Callout/formFields.tsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import type { FormField as formFieldType } from '@guardian/apps-rendering-api-models/formField';
+import type { ArticleFormat } from '@guardian/libs';
+import { remSpace, textSans } from '@guardian/source-foundations';
+import {
+	Option,
+	Select,
+	TextArea,
+	TextInput,
+} from '@guardian/source-react-components';
+import { FileInput } from '@guardian/source-react-components-development-kitchen';
```
[the remark this answers is not in the document]
Comment (sophie-macmillan) [reply apps-rendering/src/components/Callout/formFields.tsx:11]: I'll check the font and the outline in source. 

> Is it intended that once a file has been added, it only shows the 'change file' button? What if you want to remove the file and not upload a new one?
When the field isn't optional yes, you should be able to change the file but not remove it entirely as you can't submit without it. When the field is optional you see the 'change file' button as well as the 'remove button'

[the remark this answers is not in the document]
Comment (tkgnm) [reply apps-rendering/src/components/Callout/formFields.tsx:11]: Ah okay so the file upload was mandatory in that storybook example. Got it. Makes more sense now, but I still find sort of UX  a bit jarring and confusing as a user (it certainly confused me!). I think you should still be able to remove it, but just on submit it tells you that you need to upload a file. Have Ben/Bella said that's how they want it?

[the remark this answers is not in the document]
Comment (tkgnm) [reply apps-rendering/src/components/Callout/formFields.tsx:11]: Sorry - and one more thing on the file upload, I think we're meant to display 'No file chosen' when nothing is currently uploaded?

[the remark this answers is not in the document]
Comment (sophie-macmillan) [reply apps-rendering/src/components/Callout/formFields.tsx:11]: I'll check in with them on the UX for the File Input. The good news is that its in source now, so I can just update it there and leave this PR as is

File: apps-rendering/src/client/callouts.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+function callouts(): void {
+	const calloutForms = Array.from(document.querySelectorAll('.js-callout'));
+	if (!calloutForms.length) {
+		return;
+	}
+	calloutForms.forEach(setup);
+}
+
+export { callouts };
```
Comment (abeddow91) [inline apps-rendering/src/client/callouts.ts:203]: if you're looking to slim down this pr, maybe this could be pulled into its own pr?

File: apps-rendering/src/components/Callout/calloutForm.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		return Object.keys(errors).length === 0;
+	};
+
+	const onSubmit = async (formData: FormDataType): Promise<void> => {
+		// Reset error for new submission attempt
+		setSubmissionError('');
+		const isValid = validateForm();
+		if (!isValid) return;
+
+		// need to add prefix `field_` to all keys in form
```
Comment (abeddow91) [inline apps-rendering/src/components/Callout/calloutForm.tsx:88]: more for curiosity but how come we need this prefix?

File: apps-rendering/src/components/Callout/deadlineDate.test.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+		expect(getDeadlineText(date1, date2)).toBe('Open for 1 more day');
+	});
+	test('Is today if if deadline date is <24 hours', () => {
+		const date1 = new Date(2022, 11, 1, 0, 0, 0);
+		const date2 = new Date(2022, 11, 1, 2, 0, 0);
+
+		expect(getDeadlineText(date1, date2)).toBe('Closing today');
+	});
+});
```
Comment (abeddow91) [inline apps-rendering/src/components/Callout/deadlineDate.test.tsx:28]: again, great to see tests :D

File: apps-rendering/src/components/Callout/deadlineDate.test.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+		expect(getDeadlineText(date1, date2)).toBeUndefined();
+	});
+	test('Rounds days if deadline date is < 30 days away', () => {
+		const date1 = new Date(2022, 11, 1, 0, 0, 0);
+		const date2 = new Date(2022, 11, 8, 0, 0, 0);
+
+		expect(getDeadlineText(date1, date2)).toBe('Open for 7 more days');
+	});
+	test('Is singular if if deadline date is 1 day away', () => {
```
Comment (abeddow91) [inline apps-rendering/src/components/Callout/deadlineDate.test.tsx:16]: nit 
```suggestion
	test('Is singular if deadline date is 1 day away', () => {
```

File: apps-rendering/src/components/Callout/deadlineDate.test.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		const date1 = new Date(2022, 11, 1, 0, 0, 0);
+		const date2 = new Date(2022, 11, 2, 2, 0, 0);
+
+		expect(getDeadlineText(date1, date2)).toBe('Open for 1 more day');
+	});
+	test('Is today if if deadline date is <24 hours', () => {
```
Comment (abeddow91) [inline apps-rendering/src/components/Callout/deadlineDate.test.tsx:22]: nit
```suggestion
	test('Is today if deadline date is <24 hours', () => {
```

File: apps-rendering/src/components/Callout/index.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+	if (!isActive && isNonCollapsible) {
+		return (
+			<Highlight>
+				This form has been deactivated and is closed to any further
+				submissions.
+			</Highlight>
+		);
+	} else if (!isActive && !isNonCollapsible) {
+		return <></>;
```
Comment (abeddow91) [inline apps-rendering/src/components/Callout/index.tsx:44]: If the callout is not active, how come we don't return the highlight again here?

Also would it be possible to return null here or do we need the empty tags?

File: apps-rendering/src/components/Callout/calloutBlock.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+} from './styles';
+
+export interface CalloutBlockProps {
+	formId: number;
+	heading: string;
+	name: string;
+	formFields: FormField[];
+	format: ArticleFormat;
+	description?: DocumentFragment;
+	isTabbable?: boolean;
```
[the remark this answers is not in the document]
Comment (sophie-macmillan) [reply apps-rendering/src/components/Callout/calloutBlock.tsx:24]: We can! good spot!

File: apps-rendering/src/components/Callout/calloutForm.tsx (hunk 3 above)
Comment (sophie-macmillan) [reply apps-rendering/src/components/Callout/calloutForm.tsx:88]: Its a good question, its required by the [formstack api ](https://developers.formstack.com/reference/form-id-submission-post). I'll update the comment here to be more informative.

File: apps-rendering/src/components/Callout/index.tsx (hunk 7 above)
Comment (sophie-macmillan) [reply apps-rendering/src/components/Callout/index.tsx:44]: It the callout is collapsable then its an embedded callout, and after talking with Ben and Bella we decided it would be better to not show anything rather than an embedded highlight about a callout thats closed.

I'll see if null works though!

Comment (abeddow91) [reply apps-rendering/src/components/Callout/index.tsx:44]: nice, thanks for the explainer :)

**Comment by sophie-macmillan** [comment]
Closing as I'll split this large PR into smaller ones
